File: src/patterns/behavioral/iterator/iterator.go
package main

import "fmt"

//* Iterator

type Iterator[T any] interface {
	Next() T
	Current() T
	Valid() bool
	Rewind() 
	Key() int
}

//* ConcreteIterator

type NumberIterator struct {
	position int
	collection *NumberCollection
	reverse bool
}

func NewNumberIterator(collection *NumberCollection, reverse ...bool) *NumberIterator {
	if len(reverse) > 0 {
		return &NumberIterator{position: collection.GetLength()-1, collection:collection, reverse: reverse[0]}
	}
	
	return &NumberIterator{position:0, collection:collection, reverse:false}
}

func (i *NumberIterator) Next() int {
	item := i.collection.GetElements()[i.position]
	
	if i.reverse {
		i.position += -1
	} else if !i.reverse {
		i.position += 1
	}

	return item
}

func (i *NumberIterator) Current() int {
	return i.collection.GetElements()[i.position]
}

func (i *NumberIterator) Valid() bool {
	if(i.reverse) {
		return i.position >= 0
	} else {
		return i.position < i.collection.GetLength()
	}
}

func (i *NumberIterator) Rewind() {
	if(i.reverse){
		i.position = i.collection.GetLength() - 1 
	} else {
		i.position = 0
	}
}

func (i *NumberIterator) Key() int {
	return i.position
}

//* Collection

type INumberCollection interface {
	GetIterator() Iterator[int]
	GetReverseIterator() Iterator[int]
	GetLength() int
	GetElements() []int
	AddElement(el int) 
}

//* ConcreteCollection

type NumberCollection struct {
	items []int
}

func (c *NumberCollection) AddElement(el int) {
	c.items = append(c.items, el)
}

func (c *NumberCollection) GetElements() []int {
	return c.items
}

func (c *NumberCollection) GetLength() int {
	return len(c.items)
}

func (c *NumberCollection) GetIterator() Iterator[int] {
	return NewNumberIterator(c)
}

func (c *NumberCollection) GetReverseIterator() Iterator[int] {
	return NewNumberIterator(c, true)
}

//* Usage

func main(){
	numCollection := &NumberCollection{}
	numCollection.AddElement(1)
	numCollection.AddElement(2)

	iterator := numCollection.GetReverseIterator()

	for iterator.Valid() {
		fmt.Println(iterator.Next())
	}
}

File: src/patterns/behavioral/iterator/iterator.ts
//* Iterator

interface IteratorInterface<T> {
  nextItem(): T;
  current(): T;
  valid(): boolean;
  key(): number;
  rewind(): void;
}

//* Concrete Iterator

class StringIterator implements IteratorInterface<string> {
  private collection: StringCollection;
  private reverse: boolean = false;
  private position: number = 0;

  constructor(collection: StringCollection, reverse: boolean = false) {
    this.collection = collection;
    this.reverse = reverse;
    this.rewind();
  }

  nextItem(): string {
    const item = this.collection.getItems()[this.position];
    this.position += this.reverse ? -1 : 1;
    return item;
  }

  current(): string {
    return this.collection.getItems()[this.position];
  }

  valid(): boolean {
    if (this.reverse) {
      return this.position >= 0;
    }

    return this.position < this.collection.getLength();
  }

  rewind(): void {
    this.reverse ? (this.position = this.collection.getLength() - 1) : 0;
  }

  key(): number {
    return this.position;
  }
}

//* Collection

interface ICollection<T> {
  getIterator(): IteratorInterface<T>;
  getItems(): T[];
  getLength(): number;
  addItem(item: T): void;
}

//* ConcreteCollection

class StringCollection implements ICollection<string> {
  private items: string[] = [];

  getIterator() {
    return new StringIterator(this);
  }

  getReverseIterator() {
    return new StringIterator(this, true);
  }

  getItems(): string[] {
    return this.items;
  }

  addItem(item: string): void {
    this.items.push(item);
  }

  getLength(): number {
    return this.items.length;
  }
}

//* Usage

const stringCollection = new StringCollection();
stringCollection.addItem("cock");
stringCollection.addItem("and balls");

const iterator = stringCollection.getReverseIterator();
while (iterator.valid()) {
  console.log(iterator.nextItem());
}
